benches/src/streaming.rs

```diff
@@ -0,0 +1,286 @@
+    tx_progress: ProgressBar,
+    barrier: TokioMutex<()>,
+    msg_rx: AtomicU64,
+    msg_tx: AtomicU64,
+    receive_duration: Mutex<Option<Duration>>,
+}
+
+impl Context {
+    fn total(&self) -> u64 {
+        self.args.count * self.args.senders
+    }
+}
+
+const RLIMIT: u64 = 4096;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt()
+        // filter spans/events with level TRACE or higher.
+        // .with_max_level(Level::TRACE)
+        .with_env_filter(EnvFilter::new("streaming=trace"))
+        // build but do not install the subscriber.
+        .init();
+
+    let args = Args::parse();
+
+    info!("Temporarily increasing the file descriptor limit to {RLIMIT}");
+    setrlimit(Resource::NOFILE, RLIMIT, RLIMIT).expect("Failed to set file descriptor limit");
+
+    let num_msgs = args.senders * args.count;
+    let app = App {
+        ctx: Arc::new(Context {
+            args,
+            tx_progress: ProgressBar::new(num_msgs),
+            barrier: TokioMutex::default(),
+            msg_rx: AtomicU64::default(),
+            msg_tx: AtomicU64::default(),
+            receive_duration: Mutex::default(),
+        }),
+    };
+
+    let (fut, monitor_ready) = {
+        let _barrier = app.ctx.barrier.lock().await;
+        let (inbox_id, ready) = setup_monitor(app.ctx.clone()).await?;
+        info!("Receiver inbox_id: {inbox_id}");
+        let fut = setup_send_messages(inbox_id.clone(), &app.ctx).await?;
+
+        // Sleep to allow tx to send welcomes
+        sleep(Duration::from_secs(1)).await;
+
+        (fut, ready)
+    };
+
+    // Wait for the monitor thread to notify that the stream is ready.
+    monitor_ready.notified().await;
+
+    info!("Sending messages...");
+    let start = Instant::now();
+    fut.await;
+    let elapsed = start.elapsed();
+    app.ctx.tx_progress.finish();
+
+    let _ = app.ctx.barrier.lock().await;
+
+    let sent = app.ctx.msg_tx.load(Ordering::SeqCst) as i64;
+    let senders = app.ctx.args.senders;
+    let received = app.ctx.msg_rx.load(Ordering::SeqCst) as i64;
+    let dropped = sent - received;
+    let tx_elapsed = elapsed.as_secs_f32();
+    let tx_rate = sent as f32 / tx_elapsed;
+    let mut rx_elapsed = None;
+    let mut rx_rate = None;
+    if let Some(rx_duration) = *app.ctx.receive_duration.lock() {
+        let elapsed = rx_duration.as_secs_f32();
+        rx_elapsed = Some(elapsed);
+        rx_rate = Some(received as f32 / elapsed);
+    }
+
+    let rx_elapsed = rx_elapsed
+        .map(|rx| rx.to_string())
+        .unwrap_or("Unknown".to_string());
+    let rx_rate = rx_rate
+        .map(|rx| rx.to_string())
+        .unwrap_or("Unknown".to_string());
+
+    info!(
+        "\nREPORT:\n\
+        {sent} messages sent across {senders} senders,\n\
+        {received} messages received ({dropped} dropped)\n\
+        rx time: {rx_elapsed} seconds ({rx_rate} msgs/s)\n\
+        tx time: {tx_elapsed} seconds ({tx_rate} msgs/s)",
+    );
```

Comment:
I plan to switch thi to a criterion report when the rest of the benches are moved over.